lib/chef/provisioning/fog_driver/providers/xenserver.rb

```diff
@@ -0,0 +1,209 @@
+#fog:XenServer:<XenServer IP>
+class Chef
+  module Provisioning
+    module FogDriver
+      module Providers
+        class XenServer < FogDriver::Driver
+
+          Driver.register_provider_class('XenServer', FogDriver::Providers::XenServer)
+
+          def creator
+            compute_options[:xenserver_username]
+          end
+
+          def self.compute_options_for(provider, id, config)
+            new_compute_options = {}
+            new_compute_options[:provider] = provider
+            new_config = { driver_options: { compute_options: new_compute_options } }
+            new_defaults = {
+                driver_options: { compute_options: {} },
+                machine_options: { bootstrap_options: { affinity: id } }
+            }
+            result = Cheffish::MergedConfig.new(new_config, config, new_defaults)
+
+            new_compute_options[:xenserver_url] = id if id && id != ''
+            credential = Fog.credentials
+
+            new_compute_options[:xenserver_username] ||= credential[:xenserver_username]
+            new_compute_options[:xenserver_password] ||= credential[:xenserver_password]
+            new_compute_options[:xenserver_url] ||= credential[:xenserver_url]
+            new_compute_options[:xenserver_timeout] ||= 300
+            new_compute_options[:xenserver_redirect_to_master] ||= true
+
+            id = result[:driver_options][:compute_options][:xenserver_url]
+
+            [result, id]
+          end
+
+          def server_for(machine_spec)
+            if machine_spec.location
+              compute.servers.get(compute.get_by_uuid(machine_spec.location['server_id'], 'VM'))
+            else
+              nil
+            end
+          end
+
+          def servers_for(machine_specs)
+            result = {}
+            machine_specs.each do |machine_spec|
+              if machine_spec.location
+                if machine_spec.location['driver_url'] != driver_url
+                  raise "Switching a machine's driver from #{machine_spec.location['driver_url']} to #{driver_url} for is not currently supported!  Use machine :destroy and then re-create the machine on the new driver."
+                end
+                result[machine_spec] = compute.servers.get(compute.get_by_uuid(machine_spec.location['server_id'], 'VM'))
+              else
+                result[machine_spec] = nil
+              end
+            end
+            result
+          end
+          
+          def create_many_servers(num_servers, bootstrap_options, parallelizer)
+            parallelizer.parallelize(1.upto(num_servers)) do |i|
```

Comment:
An issue users have had in the past is hitting API limits.  If that will be an issue accessing the XenServer provisioner, could you change `num_servers` to `[num_servers, 5].min`?